pkg/kms/hashivault/hashivault.go

```diff
@@ -0,0 +1,251 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hashivault
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"regexp"
+
+	vault "github.com/hashicorp/vault/api"
+	"github.com/pkg/errors"
+
+	sigkms "github.com/sigstore/sigstore/pkg/kms"
+)
+
+func init() {
+	sigkms.ProvidersMux().AddProvider(ReferenceScheme, initProvider)
+}
+
+type KMS struct {
+	client  *vault.Client
+	keyPath string
+}
+
+var (
+	ErrKMSReference = errors.New("kms specification should be in the format hashivault://<key>")
+
+	re = regexp.MustCompile(`^hashivault://(?P<path>\w(([\w-.]+)?\w)?)$`)
```

Comment:
Do you think it would make sense to embed the addr/port in here to rather than env vars?

So maybe the path would be `hashivault://HASHIVAULT_ADDDR:HASHIVAULT_PORT/keypath`

I'm not sure what's most common for people here.

Reply:
I thought about that, but both looking at both Pulumi and go-cloud, they just do a simple key path and assume the appropriate env variables. I think this is related to the unique model and security constraints of Vault where you may not necessarily want to "hardcode" the Vault Address.

Reply:
Makes sense, thanks!